File: android/app/src/main/java/io/talise/app/feature/send/SendFailureView.kt
package io.talise.app.feature.send

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.PriorityHigh
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import io.talise.app.ui.theme.TaliseColors
import io.talise.app.ui.theme.TaliseType

/**
 * Terminal failure step — iOS `SendFailureView`. Reached when the pipeline
 * throws, including server 4xx rejections and transport errors. The success
 * screen is gated on a non-empty digest; every other outcome routes here.
 */
@Composable
fun SendFailureView(
    draft: SendDraft,
    onTryAgain: () -> Unit,
    onDone: () -> Unit,
) {
    Column(Modifier.fillMaxSize().background(TaliseColors.bg)) {
        Spacer(Modifier.weight(1f))

        Column(
            Modifier.fillMaxWidth(),
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(22.dp),
        ) {
            Box(
                Modifier.size(96.dp).background(TaliseColors.danger.copy(alpha = 0.15f), CircleShape),
                contentAlignment = Alignment.Center,
            ) {
                Icon(
                    Icons.Filled.PriorityHigh,
                    contentDescription = null,
                    tint = TaliseColors.danger,
                    modifier = Modifier.size(36.dp),
                )
            }

            Column(
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                Text(
                    "Send failed",
                    style = TaliseType.heading(34.sp, FontWeight.Medium),
                    letterSpacing = (-1).sp,
                    color = TaliseColors.fg,
                )
                Text(
                    "No funds moved. You can try again or close this.",
                    style = TaliseType.body(14.sp, FontWeight.Light),
                    color = TaliseColors.fgMuted,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(horizontal = 32.dp),
                )
            }

            val err = draft.errorMessage
            if (!err.isNullOrEmpty()) {
                Text(
                    err,
                    style = TaliseType.body(13.sp, FontWeight.Light),
                    color = TaliseColors.fgMuted,
                    textAlign = TextAlign.Center,
                    modifier = Modifier.padding(horizontal = 32.dp).padding(top = 4.dp),
                )
            }
        }

        Spacer(Modifier.weight(1f))

        Column(
            Modifier.padding(horizontal = 24.dp).padding(bottom = 18.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            SolidCapsuleButton(title = "Try again", onClick = onTryAgain)
            GlassCapsuleButton(title = "Done", onClick = onDone)
        }
    }
}

File: ios/Talise/Features/Send/SendFailureView.swift
import SwiftUI

/// Terminal failure step for the Send flow. Reached when sponsor-prepare,
/// sponsor-execute, or gasless-submit throws — including server 4xx
/// rejections like `ACCUMULATOR_UNDERFUNDED` and transport errors.
///
/// This screen is the deliberate replacement for the historical bug where
/// the green success checkmark rendered with an error string stacked
/// beneath it. The success screen is now gated on a non-empty digest;
/// every other outcome routes here.
struct SendFailureView: View {
    @Bindable var draft: SendDraft
    var onTryAgain: () -> Void
    var onDone: () -> Void

    var body: some View {
        VStack(spacing: 0) {
            Spacer()

            VStack(spacing: 22) {
                ZStack {
                    Circle()
                        .fill(TaliseColor.danger.opacity(0.15))
                        .frame(width: 96, height: 96)
                    Image(systemName: "exclamationmark")
                        .font(.system(size: 36, weight: .semibold))
                        .foregroundStyle(TaliseColor.danger)
                }

                VStack(spacing: 8) {
                    Text("Send failed")
                        .font(TaliseFont.heading(34, weight: .medium))
                        .kerning(-1)
                        .foregroundStyle(TaliseColor.fg)
                    // Deliberately does NOT assert "no funds moved". This screen
                    // is shown for any failure, including a transport timeout
                    // that can fire AFTER the transaction was broadcast — in
                    // which case the money did move, and telling the user
                    // otherwise invites them to send a second time.
                    Text("Check your activity before trying again — if the send went through, it will show there.")
                        .font(TaliseFont.body(14, weight: .light))
                        .foregroundStyle(TaliseColor.fgMuted)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 32)
                }

                if let err = draft.errorMessage, !err.isEmpty {
                    Text(err)
                        .font(TaliseFont.body(13, weight: .light))
                        .foregroundStyle(TaliseColor.fgMuted)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, 32)
                        .padding(.top, 4)
                }
            }

            Spacer()

            VStack(spacing: 10) {
                Button(action: onTryAgain) {
                    Text("Try again")
                        .font(TaliseFont.heading(16, weight: .medium))
                        .foregroundStyle(TaliseColor.bg)
                        .frame(maxWidth: .infinity)
                        .frame(height: 56)
                        .background(TaliseColor.fg)
                        .clipShape(Capsule())
                }
                Button(action: onDone) {
                    Text("Done")
                        .font(TaliseFont.heading(16, weight: .medium))
                        .foregroundStyle(TaliseColor.fg)
                        .frame(maxWidth: .infinity)
                        .frame(height: 56)
                        .glassCapsule()
                }
            }
            .padding(.horizontal, 24)
            .padding(.bottom, 18)
        }
        .taliseScreenBackground()
        .toolbar(.hidden, for: .navigationBar)
    }
}
